### PR TITLE
Add eval_logits_to_cpu

### PR DESCRIPTION
**Pull request by EkhiAzur**
# What does this PR do?

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by EkhiAzur** [comment]
Yes sorry, I missed that part.

With large generative models during evaluation if compute metrics function is provided, OOM errors can occur. This error happens because predictions logits are save in GPU during evaluation. Eval_accumulation_steps is not the best option to avoid this error because it really slows the run. To avoid that, I added a parameter to decide if logits are saved in the GPU or in CPU.

Thanks for you attention

**Comment by amyeroberts** [comment]
@EkhiAzur Thanks for explaining! Trainer already has very many arguments and so we want to be cautious about each one we add. The one being proposed here is highly specific and I suspect not the only (or best) way to tackle OOM when using trainer and accelerate. I'll let @muellerzr  and @pacman100 chime in on whether this should be added and possible alternatives

**Comment by muellerzr** [comment]
@EkhiAzur I can't see a reason why things would break if we just did this all the time for the logits instead. Since it's after `gather` this should be safe

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -3225,6 +3225,8 @@ def evaluation_loop(
                 if self.preprocess_logits_for_metrics is not None:
                     logits = self.preprocess_logits_for_metrics(logits, labels)
                 logits = self.accelerator.gather_for_metrics((logits))
+                if args.eval_logits_to_cpu:
+                    logits = logits.cpu()
```
Comment (pacman100) [inline src/transformers/trainer.py:3229]: this would fail if the logits are a tuple of tensors which can be the case as per below line in prediction step:

https://github.com/huggingface/transformers/blob/e66984f9952641e176521a7ef3faf2d5fa3c608b/src/transformers/trainer.py#L3440

**Review by pacman100 — COMMENTED** [review]
Hello, thank you for the PR with the aim to reduce GPU memory during evaluation, but I think this is incorrect as per my comment.

File: src/transformers/trainer.py (hunk 1 above)
Comment (muellerzr) [reply src/transformers/trainer.py:3229]: Agreed. Can we bring in `to_device` from `accelerate` instead maybe? (I didn't see a similar util in transformers, but safe to say we can use it here)

**Comment by EkhiAzur** [comment]
Hello, I realized that using `preprocess_logits_for_metrics` to reduce logits shape is better than the PR change. Sorry about that. 

Here is my solution if anyone needs it:

`def preprocess_logits_for_metrics(logits, labels):
    if type(logits)==tuple:
        logits = logits[0]
    logits = logits.argmax(axis=-1)
    return logits` 

And, then pass it as an argument to the Trainer:
`
trainer = Trainer(
    model=model,
    args=trainingArgs,
    preprocess_logits_for_metrics = preprocess_logits_for_metrics,
    ....
)
`

**Comment by muellerzr** [comment]
@EkhiAzur can we close this then? :)

**Comment by EkhiAzur** [comment]
Yes, thanks for your attention and help!
